### ISSUE TITLE
Java Records should automatically be flagged as @Immutable

### ISSUE THREAD

**Issue by octylFractal** [google/error-prone#5420]
Records are frequently going to be immutable types, and I think it's kind of annoying to have to annotate them with `@Immutable` when the modifier should be enough of an indication.

I am aware that records may contain mutable types, but my idea was that the `Immutable` checker would also validate that a record does not contain those either, which should be straightforward.

I encountered this because I have an enum which contains a `Vector3`:
```java
record Vector3(double x, double y, double z) {}

enum Direction {
    NORTH(new Vector3(0, 0, -1)),
    EAST(new Vector3(1, 0, 0)),
    SOUTH(new Vector3(0, 0, 1)),
    WEST(new Vector3(-1, 0, 0));

    private final Vector3 direction;

    Direction(Vector3 vector) {
        this.direction = vector;
    }
}
```

**Comment by Tharanishwaran** [comment]
Hi! I'd like to work on this issue I contribute to CNCF and Spring projects and can implement automatic @Immutable treatment for records with proper validation and tests

**Comment by graememorgan** [comment]
This request has come up internally as well. :) A few points:

We can't tell whether a given referenced type is immutable at compile-time, given we may not have access to its `private` fields (if it was compiled separately). So, we'd be limited to considering `record`s as implicitly immutable iff all their fields are.

This is a _bit_ similar to `enum`s which we regard as implicitly immutable, except that we _unconditionally_ consider them to be immutable when referencing their types (even if they're not). Checking a `record`-of-`record`s would require recursing down the hierarchy of `record`s to check all the leaf fields. That might be expensive.

It might also be a bit fragile. It would mean that adding a mutable field to a previously-implicitly-immutable `record` might break compilation elsewhere, which is much less obvious than breaking compilation by removing an `@Immutable` annotation.

**Comment by graememorgan** [comment]
Another thought: you could make a very similar argument for `AutoValue`s. I don't think we'd want the behaviour of the two to be different.

(The slight difference is that they're not `final` and you _can_ override them, although we do have a build error in place to help prevent that.)
